cleverhans/experimental/certification/nn.py

```diff
@@ -0,0 +1,157 @@
+    self.final_weights = tf.convert_to_tensor(
+        net_weights[self.num_hidden_layers], dtype=tf.float32)
+    self.final_bias = tf.convert_to_tensor(
+        net_biases[self.num_hidden_layers], dtype=tf.float32)
+
+  def forward_pass(self, vector, layer_index, is_transpose=False, is_abs=False):
+    """Performs forward pass through the layer weights at layer_index.
+
+    Args:
+      vector: vector that has to be passed through in forward pass
+      layer_index: index of the layer
+      is_transpose: whether the weights of the layer have to be transposed
+      is_abs: whether to take the absolute value of the weights
+
+    Returns:
+      tensor that corresponds to the forward pass through the layer
+    Raises:
+      ValueError: if the layer_index is negative or more than num hidden layers
+    """
+    if (layer_index < 0 or layer_index > self.num_hidden_layers):
+      raise ValueError('Invalid layer index')
+
+    if (self.layer_types[layer_index] == 'ff' or
+        self.layer_types[layer_index] == 'ff_relu'):
+      if is_transpose:
+        return tf.matmul(tf.transpose(self.weights[layer_index]), vector)
+      elif is_abs:
+        return tf.matmul(tf.abs(self.weights[layer_index]), vector)
+      else:
+        return tf.matmul(self.weights[layer_index], vector)
+
+    raise NotImplementedError('Unsupported layer type: {0}'.format(
+        self.layer_types[layer_index]))
+
+def load_network_from_checkpoint(checkpoint, model_json):
+  """Function to read the weights from checkpoint based on json description.
+
+    Args:
+      checkpoint: tensorflow checkpoint with trained model to
+        verify
+      model_json: path of json file with model description of
+        the network list of dictionary items for each layer
+        containing 'type', 'weight_var', 'bias_var' and
+        'is_transpose' 'type'is one of {'ff', 'ff_relu' or
+        'conv'}; 'weight_var' is the name of tf variable for
+        weights of layer i; 'bias_var' is the name of tf
+        variable for bias of layer i; 'is_transpose' is set to
+        True if the weights have to be transposed as per
+        convention Note that last layer is always feedforward
+        (verification operates at the layer below the final
+        softmax for more numerical stability)
+
+    Returns:
+      net_weights: list of numpy matrices of weights of each
+      layer
+    [convention: x[i+1] = W[i] x[i]
+    net_biases: list of numpy arrays of biases of each layer
+    net_layer_types: type of each layer ['ff' or 'ff_relu' or 'ff_conv'
+      or 'ff_conv_relu']
+      'ff': Simple feedforward layer with no activations
+      'ff_relu': Simple feedforward layer with ReLU activations
+      'ff_conv': Convolution layer with no activation
+      'ff_conv_relu': Convolution layer with ReLU activation
+
+    Raises:
+      ValueError: If layer_types are invalid or variable names
+      not found in
```

Comment:
please fix indentation (add two spaces), also please combine this and next line together (i.e. `not found in checkpoint` in the same line)